### ISSUE TITLE
sphinx-doc crashes in html output

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1568]
While building the following assertion shows up without more explanation (see below).
Building text output of the same document succeeds.

```
#!text

# Sphinx version: 1.2.2
# Python version: 2.7.8
# Docutils version: 0.12 release
# Jinja2 version: 2.7.3
# Loaded extensions:
#   sphinx.ext.autodoc from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/autodoc.pyc
#   sphinx.ext.doctest from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/doctest.pyc
#   sphinx.ext.viewcode from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/viewcode.pyc
#   sphinx.ext.todo from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/todo.pyc
#   sphinx.ext.ifconfig from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/ifconfig.pyc
#   sphinx.ext.oldcmarkup from /project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/ext/oldcmarkup.pyc
Traceback (most recent call last):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/cmdline.py", line 254, in main
    app.build(force_all, filenames)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/application.py", line 212, in build
    self.builder.build_update()
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 214, in build_update
    'out of date' &#37; len(to_build))
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 276, in build
    self.write(docnames, list(updated_docnames), method)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 320, in write
    self._write_serial(sorted(docnames), warnings)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 333, in _write_serial
    self.write_doc(docname, doctree)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/builders/html.py", line 433, in write_doc
    self.docwriter.write(doctree, destination)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/writers/__init__.py", line 80, in write
    self.translate()
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/writers/html.py", line 51, in translate
    self.document.walkabout(visitor)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 166, in walkabout
    visitor.dispatch_visit(self)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/docutils/nodes.py", line 1882, in dispatch_visit
    return method(node)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/writers/html.py", line 204, in visit_reference
    assert len(node) == 1 and isinstance(node[0], nodes.image)
AssertionError

```

Excerpt from "sphinxbuild" with increased verbosity:

```
#!text

[...]
preparing documents... done
writing output... [ 14&#37;] appendix
[app] emitting event: 'doctree-resolved'(<document: <highlightlang...><substitution_definition "setupenv"... ...>, u'appendix')
[app] emitting event: 'html-page-context'(u'appendix', 'page.html', {'file_suffix': '.html', 'has_source': True, 'show_sphinx': True, 'last_u
writing output... [ 28&#37;] conventions
[app] emitting event: 'doctree-resolved'(<document: <highlightlang...><substitution_definition "setupenv"... ...>, u'conventions')
[app] emitting event: 'html-page-context'(u'conventions', 'page.html', {'file_suffix': '.html', 'has_source': True, 'show_sphinx': True, 'las
writing output... [ 42&#37;] dev_workflows
[app] emitting event: 'doctree-resolved'(<document: <highlightlang...><section "development process workflow ...>, u'dev_workflows')
[app] emitting event: 'build-finished'(AssertionError(),)

Traceback (most recent call last):
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/cmdline.py", line 254, in main
    app.build(force_all, filenames)
  File "/project/scm/opt/python2.7/lib/python2.7/site-packages/sphinx/application.py", line 212, in build
    self.builder.build_update()

[... and the traceback dump continues ... ]


```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1568
- Originally reported by: Waldek M.
- Originally created at: 2014-09-17T11:58:05.525

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 10:03:02+00:00_

Thanks for the report. Can you attach the input .rst document that led to the assertion? (If you build with `sphinx-build -v` or `make html SPHINXOPTS=-v` you can see better which document is being translated at the time).

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 10:04:01+00:00_

Ah, our edits crossed. So, can you attach the `dev_workflows` file? (Or is the Sphinx project available online?)

**Comment by shimizukawa** [comment]
_From Waldek M. on 2014-09-17 10:25:11+00:00_

I'd like to, but my employer might not be happy ;-)
I'll try to extract a test sample (might take a while).

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 10:30:01+00:00_

Thanks. What could also be helpful (if you're familiar with Python and pdb) is to run Sphinx with `-P` and see what `node` and `node[0]` are in the last stackframe.

**Comment by shimizukawa** [comment]
_From Waldek M. on 2014-09-17 11:29:03+00:00_

It's a bit rusty, but here's what I was able to gather.

```
#!text

(Pdb) from pprint import pprint
(Pdb) p node[0]
<#text: 'ccmerge'>
(Pdb) p node[0].__class__
<class 'docutils.nodes.Text'>
(Pdb) pprint(vars(node[0]))
{'parent': <reference: <#text: 'ccmerge'>>, 'rawsource': ''}
(Pdb) p node
<reference: <#text: 'ccmerge'>>
(Pdb) pprint (node)
<reference: <#text: 'ccmerge'>>
(Pdb) pprint(vars(node))
{'attributes': {'backrefs': [],
                'classes': [],
                'dupnames': [],
                'ids': [],
                'name': u'ccmerge',
                'names': [],
                'refuri': u'http://xyz....'},
 'children': [<#text: 'ccmerge'>],
 'parent': <centered: <#text: 'Repeat the '><reference...><#text: ' command un ...>,
 'rawsource': u'`ccmerge`_',
 'resolved': 1,
 'tagname': 'reference'}
(Pdb) p len(node)
1
```

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 11:35:08+00:00_

Thanks, that is exactly what I need! Please check if the following patch fixes the issue:

```
diff -r 3ae11a0d92af sphinx/addnodes.py
--- a/sphinx/addnodes.py        Wed Sep 17 12:26:40 2014 +0200
+++ b/sphinx/addnodes.py        Wed Sep 17 13:34:11 2014 +0200
@@ -107,7 +107,7 @@ class index(nodes.Invisible, nodes.Inlin
     *entrytype* is one of "single", "pair", "double", "triple".
     """

-class centered(nodes.Part, nodes.Element):
+class centered(nodes.Part, nodes.TextElement):
     """Deprecated."""

 class acks(nodes.Element):
```

**Comment by shimizukawa** [comment]
_From Waldek M. on 2014-09-17 11:51:13+00:00_

Just started to realize it might be somehow related to the nasty "centered" directive and here you are with the solution :-)

Works like a charm. 

Now I'm not only amazed with sphinx itself but also stunned by your speed.
Thank you very, very much!

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 12:14:21+00:00_

Closes #1568: since "centered" directly contains text children, it should be a TextElement.

→ <<cset d2b955b7f9421f431a8e48f613651468d33e7304>>

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-09-17 12:14:24+00:00_

You're welcome!
